File: src/convex/ml/runs.ts
// Pure-TS run-scoring model for MLB game totals and run lines.
// No node/convex imports — safe to use from Convex actions and shared types.
//
// Approach:
//   1. Fit each team's offensive / defensive run factor from historical scores
//      plus a per-ballpark factor (all derived from the single game dataset).
//   2. Model each team's runs as independent Poisson random variables with
//      mean = leagueAvg * offense(home) * defense(away) * park(home).
//   3. Monte Carlo (10,000 simulations) produces predicted scores, the total
//      over/under distribution, and run-line (±1.5) cover probabilities.

import { RawGame } from "./types";

export interface RunModel {
  leagueRuns: number;
  teamOffense: Record<number, number>;
  teamDefense: Record<number, number>;
  parkFactor: Record<number, number>; // keyed by home team id (venue proxy)
}

export interface RunSimulation {
  homeScore: number; // mean runs
  awayScore: number; // mean runs
  total: number; // mean combined runs
  overProb: number; // P(total > line)
  underProb: number; // P(total < line)
  homeRunLineProb: number; // P(home wins by 2+ → covers -1.5)
  awayRunLineProb: number; // P(away loses by ≤1 or wins → covers +1.5)
}

function clamp(x: number, lo: number, hi: number): number {
  return Math.min(hi, Math.max(lo, x));
}

/** Deterministic xorshift PRNG so Monte Carlo results are reproducible. */
function makeRand(seed: number): () => number {
  let s = (seed >>> 0) || 0x12345678;
  return () => {
    s ^= s << 13;
    s >>>= 0;
    s ^= s >>> 17;
    s ^= s << 5;
    s >>>= 0;
    return s / 4294967296;
  };
}

/** Knuth's Poisson sampler. */
function poisson(lambda: number, rand: () => number): number {
  if (lambda <= 0) return 0;
  const L = Math.exp(-lambda);
  let k = 0;
  let p = 1;
  do {
    k += 1;
    p *= rand();
  } while (p > L);
  return k - 1;
}

export function fitRunModel(games: RawGame[]): RunModel {
  const teamGames = new Map<number, number>();
  const teamScored = new Map<number, number>();
  const teamAllowed = new Map<number, number>();
  const parkRuns = new Map<number, { runs: number; games: number }>();
  let totalRuns = 0;
  let totalGames = 0;

  for (const g of games) {
    if (g.winner !== "home" && g.winner !== "away") continue;
    const hs = g.home.score;
    const as = g.away.score;
    if (typeof hs !== "number" || typeof as !== "number") continue;
    const homeId = g.home.id;
    const awayId = g.away.id;
    teamGames.set(homeId, (teamGames.get(homeId) ?? 0) + 1);
    teamGames.set(awayId, (teamGames.get(awayId) ?? 0) + 1);
    teamScored.set(homeId, (teamScored.get(homeId) ?? 0) + hs);
    teamAllowed.set(homeId, (teamAllowed.get(homeId) ?? 0) + as);
    teamScored.set(awayId, (teamScored.get(awayId) ?? 0) + as);
    teamAllowed.set(awayId, (teamAllowed.get(awayId) ?? 0) + hs);
    const pr = parkRuns.get(homeId) ?? { runs: 0, games: 0 };
    parkRuns.set(homeId, { runs: pr.runs + hs + as, games: pr.games + 1 });
    totalRuns += hs + as;
    totalGames += 1;
  }

  const leagueRuns = totalGames > 0 ? totalRuns / (2 * totalGames) : 4.4;

  const teamOffense: Record<number, number> = {};
  const teamDefense: Record<number, number> = {};
  const parkFactor: Record<number, number> = {};
  for (const id of teamGames.keys()) {
    const g = teamGames.get(id) ?? 1;
    teamOffense[id] = clamp((teamScored.get(id) ?? 0) / g / leagueRuns, 0.7, 1.3);
    teamDefense[id] = clamp((teamAllowed.get(id) ?? 0) / g / leagueRuns, 0.7, 1.3);
  }
  for (const [id, p] of parkRuns) {
    parkFactor[id] = clamp(p.runs / p.games / (2 * leagueRuns), 0.85, 1.15);
  }

  return { leagueRuns, teamOffense, teamDefense, parkFactor };
}

/**
 * Monte Carlo run simulation for a matchup. `line` is the total over/under
 * reference (market total when available, otherwise the model mean total).
 */
export function simulateRuns(
  model: RunModel,
  homeId: number,
  awayId: number,
  line: number,
  trials = 10000,
): RunSimulation {
  const offense = model.teamOffense;
  const defense = model.teamDefense;
  const park = model.parkFactor;
  const parkMul = park[homeId] ?? 1;
  const lambdaHome = model.leagueRuns * (offense[homeId] ?? 1) * (defense[awayId] ?? 1) * parkMul;
  const lambdaAway = model.leagueRuns * (offense[awayId] ?? 1) * (defense[homeId] ?? 1) * parkMul;

  const rand = makeRand((homeId * 1000003 + awayId * 7919) >>> 0);
  let homeSum = 0;
  let awaySum = 0;
  let over = 0;
  let under = 0;
  let homeCover = 0;
  let awayCover = 0;
  for (let t = 0; t < trials; t++) {
    const hs = poisson(lambdaHome, rand);
    const as = poisson(lambdaAway, rand);
    homeSum += hs;
    awaySum += as;
    const total = hs + as;
    if (total > line) over += 1;
    else if (total < line) under += 1;
    const margin = hs - as;
    if (margin >= 2) homeCover += 1;
    if (margin <= 1) awayCover += 1;
  }

  return {
    homeScore: homeSum / trials,
    awayScore: awaySum / trials,
    total: (homeSum + awaySum) / trials,
    overProb: over / trials,
    underProb: under / trials,
    homeRunLineProb: homeCover / trials,
    awayRunLineProb: awayCover / trials,
  };
}

File: mlb_streamlit/engine/runs.py
"""Run-scoring model: Poisson offense/defense factors + Monte Carlo.

Faithful port of src/convex/ml/runs.ts. Each team's runs are modeled as
independent Poisson variables with mean = leagueAvg * offense * defense * park,
then a deterministic xorshift PRNG drives the Monte Carlo simulation so
results are reproducible.
"""

from __future__ import annotations

import math


def clamp(x: float, lo: float, hi: float) -> float:
    return min(hi, max(lo, x))


def make_rand(seed: int):
    """Deterministic xorshift PRNG so Monte Carlo results are reproducible."""
    s = (seed & 0xFFFFFFFF) or 0x12345678

    def rand() -> float:
        nonlocal s
        s ^= (s << 13) & 0xFFFFFFFF
        s ^= s >> 17
        s ^= (s << 5) & 0xFFFFFFFF
        s &= 0xFFFFFFFF
        return s / 4294967296.0

    return rand


def poisson(lambda_: float, rand) -> int:
    """Knuth's Poisson sampler."""
    if lambda_ <= 0:
        return 0
    L = math.exp(-lambda_)
    k = 0
    p = 1.0
    while True:
        k += 1
        p *= rand()
        if p <= L:
            break
    return k - 1


def fit_run_model(games: list[dict]) -> dict:
    """Fit league average + per-team offense/defense/park factors from scores."""
    team_games: dict[int, int] = {}
    team_scored: dict[int, float] = {}
    team_allowed: dict[int, float] = {}
    park_runs: dict[int, dict] = {}
    total_runs = 0
    total_games = 0

    for g in games:
        if g["winner"] not in ("home", "away"):
            continue
        hs = g["home"].get("score")
        as_ = g["away"].get("score")
        if not isinstance(hs, (int, float)) or not isinstance(as_, (int, float)):
            continue
        home_id = g["home"]["id"]
        away_id = g["away"]["id"]
        team_games[home_id] = team_games.get(home_id, 0) + 1
        team_games[away_id] = team_games.get(away_id, 0) + 1
        team_scored[home_id] = team_scored.get(home_id, 0) + hs
        team_allowed[home_id] = team_allowed.get(home_id, 0) + as_
        team_scored[away_id] = team_scored.get(away_id, 0) + as_
        team_allowed[away_id] = team_allowed.get(away_id, 0) + hs
        pr = park_runs.get(home_id, {"runs": 0, "games": 0})
        pr["runs"] += hs + as_
        pr["games"] += 1
        park_runs[home_id] = pr
        total_runs += hs + as_
        total_games += 1

    league_runs = total_runs / (2 * total_games) if total_games > 0 else 4.4

    team_offense: dict[int, float] = {}
    team_defense: dict[int, float] = {}
    park_factor: dict[int, float] = {}
    for tid, g_count in team_games.items():
        team_offense[tid] = clamp((team_scored.get(tid, 0) / g_count) / league_runs, 0.7, 1.3)
        team_defense[tid] = clamp((team_allowed.get(tid, 0) / g_count) / league_runs, 0.7, 1.3)
    for tid, p in park_runs.items():
        park_factor[tid] = clamp(p["runs"] / p["games"] / (2 * league_runs), 0.85, 1.15)

    return {
        "leagueRuns": league_runs,
        "teamOffense": team_offense,
        "teamDefense": team_defense,
        "parkFactor": park_factor,
    }


def expected_margin(model: dict, home_id: int, away_id: int) -> float:
    park_mul = model["parkFactor"].get(home_id, 1)
    lr = model["leagueRuns"]
    lambda_home = lr * model["teamOffense"].get(home_id, 1) * model["teamDefense"].get(away_id, 1) * park_mul
    lambda_away = lr * model["teamOffense"].get(away_id, 1) * model["teamDefense"].get(home_id, 1) * park_mul
    return lambda_home - lambda_away


def expected_total(model: dict, home_id: int, away_id: int) -> float:
    park_mul = model["parkFactor"].get(home_id, 1)
    lr = model["leagueRuns"]
    return (
        lr
        * (
            model["teamOffense"].get(home_id, 1) * model["teamDefense"].get(away_id, 1)
            + model["teamOffense"].get(away_id, 1) * model["teamDefense"].get(home_id, 1)
        )
        * park_mul
    )


def simulate_runs(
    model: dict,
    home_id: int,
    away_id: int,
    line: float,
    trials: int = 10000,
    run_line: float = 1.5,
    margin_shift: float = 0.0,
) -> dict:
    """Monte Carlo run simulation. `marginShift` reconciles scores with the
    win-probability model (total preserved)."""
    offense = model["teamOffense"]
    defense = model["teamDefense"]
    park = model["parkFactor"]
    park_mul = park.get(home_id, 1)
    base_home = model["leagueRuns"] * offense.get(home_id, 1) * defense.get(away_id, 1) * park_mul
    base_away = model["leagueRuns"] * offense.get(away_id, 1) * defense.get(home_id, 1) * park_mul
    shift = clamp(margin_shift, -min(base_home, base_away) + 0.05, min(base_home, base_away) - 0.05)
    lambda_home = base_home + shift
    lambda_away = base_away - shift
    cover_threshold = math.ceil(run_line)  # 1.5 -> 2, 2.5 -> 3

    rand = make_rand((home_id * 1000003 + away_id * 7919) & 0xFFFFFFFF)
    home_sum = 0
    away_sum = 0
    over = 0
    under = 0
    home_cover = 0
    away_cover = 0
    for _ in range(trials):
        hs = poisson(lambda_home, rand)
        as_ = poisson(lambda_away, rand)
        home_sum += hs
        away_sum += as_
        total = hs + as_
        if total > line:
            over += 1
        elif total < line:
            under += 1
        margin = hs - as_
        if margin >= cover_threshold:
            home_cover += 1
        else:
            away_cover += 1

    over_under = over + under
    return {
        "homeScore": home_sum / trials,
        "awayScore": away_sum / trials,
        "total": (home_sum + away_sum) / trials,
        "overProb": (over / over_under) if over_under > 0 else 0.5,
        "underProb": (under / over_under) if over_under > 0 else 0.5,
        "homeRunLineProb": home_cover / trials,
        "awayRunLineProb": away_cover / trials,
    }
